### ISSUE TITLE
reqd_work_group_size and non-uniform NDRanges

### ISSUE THREAD

**Issue by kpet** [KhronosGroup/OpenCL-Docs#267]
The OpenCL C 2.0 specification currently reads (§6.7.2):

>The optional __attribute__((reqd_work_group_size(X, Y, Z))) is the work-group size that must be
used as the local_work_size argument to clEnqueueNDRangeKernel. This allows the compiler to
optimize the generated code appropriately for this kernel.

The only thing the compiler can know is that the work group size will be in the [{1,1,1},reqd_work_group_size] range when non-uniform NDRanges are in use which seems to defeat the purpose of having `reqd_work_group_size`.

FWIW, the non_uniform_work_group CTS tests do exercise the case where the attibute is specified but the NDRange is such that some work groups are smaller than the value specified with `reqd_work_group_size`.

Should we require NDRange uniformity when `reqd_work_group_size` is used instead?

**Comment by bashbaug** [comment]
I agree the current behavior is a little non-intuitive, but I do believe it was intended.  For a pure OpenCL 2.0 compile, the `reqd_work_group_size` kernel attribute guarantees that `get_enqueued_local_size` will return the value specified by the attribute, but because work group sizes may be non-uniform the only guarantee for `get_local_size` is that it is less than or equal to the value specified by the attribute.

If applications also want to guarantee the value of `get_local_size`, they may do so by passing `-cl-uniform-work-group-size` to require uniform work groups.

**Comment by kpet** [comment]
Agree with all of that but the question I was asking was really: "does it make sense to allow users to express a required work group size that can't be a hard requirement?" :). I would expect the typical use for `reqd_work_group_size` to happen in kernels where either assumptions have to be made about the workgroup size or where fixing the workgroup size leads to performance improvements. `-cl-uniform-work-group-size` affects the entire program...

**Comment by bashbaug** [comment]
> I would expect the typical use for `reqd_work_group_size` to happen in kernels where either assumptions have to be made about the workgroup size or where fixing the workgroup size leads to performance improvements. `-cl-uniform-work-group-size` affects the entire program...

Thinking out loud: should there be a kernel attribute that says the work group size must be uniform for a specific kernel?  I think that could be useful for finer grain control both with and without `reqd_work_group_size`.

**Comment by kpet** [comment]
Discussed on the 2020/05/12 call:
- `reqd_work_group_size` is essentially equivalent to `work_group_size_hint` in the non-uniform case (except that the latter is not queryable from the API)
- Kernels written for OpenCL 1.2 had a strong guarantee that the value returned by `get_local_size()` was equal to the value specified as `reqd_work_group_size`. These would silently break if used with non-uniform NDRanges on an OpenCL 2.x/3.x implementation.
- Some said that they would expect enqueues of kernels that have a `reqd_work_group_size` with a non-uniform NDRange to fail. 

We already had https://gitlab.khronos.org/opencl/OpenCL-Docs/issues/130 within Khronos to discuss this.

**Comment by bashbaug** [comment]
I don't think this is quite true?

> * `reqd_work_group_size` is essentially equivalent to `work_group_size_hint` in the non-uniform case (except that the latter is not queryable from the API)

The work group size for the **uniform** work groups must be equal to the `reqd_work_group_size` in the non-uniform case, but there is no requirement that the remainder work groups must be equal to the required work group size.

Given a kernel with `reqd_work_group_size(16, 1, 1)`, I would expect:

```c
clEnqueueNDRangeKernel(gws=16, lws=16); // --> pass
clEnqueueNDRangeKernel(gws=24, lws=16); // --> pass
clEnqueueNDRangeKernel(gws=16, lws=8); // --> fail
clEnqueueNDRangeKernel(gws=8, lws=16); // --> pass
```

* The first case has one uniform work group of size 16, which matches the required work group size. 
* The second case has one uniform work group of size 16, which matches the required work group size, and one remainder work group of size 8.
* The third case has two uniform work groups of size 8, which does not match the required work group size, so this is an error.
* The last case is a little weird since there are no uniform work groups and there are only remainder work groups, but I think this should be OK too.  For this one work group, I would expect `get_enqueued_local_size()` to be `16`, and `get_local_size()` to be `8`.

**Comment by kpet** [comment]
@bashbaug Yes, agree with your comment, I really had the compiler side in mind when writing this down. I should have been more specific in my comment. The comment was really that `reqd_work_group_size` doesn't allow any more optimisation than `work_group_size_hint`.

**Comment by bashbaug** [comment]
Discussed in the June 2nd teleconference.  No conclusion, but we are trending toward a new kernel attribute that would require uniform work groups on the kernel side, similar to the `-cl-uniform-work-group-size` build option on the host side.  This attribute would be usable both with and without the required work group size kernel attribute.  When combined with the required work group size kernel attribute, it allows OpenCL C 2.0 compilers and OpenCL C 3.0 compilers that support non-uniform work groups to replace calls to `get_enqueued_local_size()` and `get_local_size()` with the required work group size.

**Comment by kpet** [comment]
Adding the OpenCL 3 Must Fix label since resolving this will involve adding new functionality.

**Comment by bashbaug** [comment]
Moving off of the OpenCL 3 board, as discussed in the July 28th teleconference.
